### PR TITLE
setClientVersion-implementation as discussed in #129

### PR DESCRIPTION
**Pull request by kimmerin**
Compared to the logger issue this is a more simple change. I've added a system-property that allows to set the client version that way as well and added tests to check the behavior. This required the setup of the config-hashtable to be in a dedicated method.

### CODE REVIEW

File: src/main/java/com/jcraft/jsch/Session.java [hunk 1]
```diff
@@ -2716,9 +2716,13 @@ public String getServerVersion(){
     return Util.byte2str(V_S);
   }
   public String getClientVersion(){
-    return Util.byte2str(V_C);
+    return V_C == null ? jsch.getClientVersion() : Util.byte2str(V_C);
   }
   public void setClientVersion(String cv){
+    V_C = null;
```
Comment (norrisjeremy) [inline src/main/java/com/jcraft/jsch/Session.java:2722]: This will lead to an NPE if a user explicitly calls `setClientVersion(null)` because the `V_C` property is dereferenced in the `connect()` & `receive_kexinit()` methods.

**Review by norrisjeremy — CHANGES_REQUESTED** [review]

File: src/main/java/com/jcraft/jsch/Session.java (hunk 1 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/Session.java:2723]: All the tests that actually show that require a Docker-container I don't have. I've changed it so that everything is using the get-method now

File: src/main/java/com/jcraft/jsch/Session.java [hunk 2]
```diff
@@ -624,7 +625,7 @@ private KeyExchange receive_kexinit(Buffer buf) throws Exception {
       throw new JSchException(e.toString(), e);
     }
 
-    kex.doInit(this, V_S, V_C, I_S, I_C);
+    kex.doInit(this, V_S, Util.str2byte(getClientVersion()), I_S, I_C);
```
Comment (norrisjeremy) [inline src/main/java/com/jcraft/jsch/Session.java:628]: If a user changed the client version string after a Session is connected and a rekeying operation occurred, their SSH connection would then break, because the client version is supposed to be idempotent.

Whilst this could already happen prior, with how you've adapted the Session class to fallback to calling an instance method in the JSch class, we would be making this scenario easier for a user to trigger, since they may naively believe they can change the client version of the JSch instance between multiple calls to `getSession()`.

I'm thinking we should avoid this scenario, by simply initializing the `V_C` property to whatever the value of `jsch.getClientVersion()` is at the time the Session instance is created and avoid trying the callbacks to `jsch.getClientVersion()` within `session.getClientVersion()`.

We simply would need to document to users that the client version of a Session instance is not affected by changes to the JSch instance after the Session is created, and that the only way to manipulate afterwards is by direct calls to `session.setClientVersion()`.

**Review by norrisjeremy — CHANGES_REQUESTED** [review]

File: src/main/java/com/jcraft/jsch/JSch.java [hunk 3]
```diff
@@ -42,7 +43,12 @@
 
   static Hashtable<String, String> config=new Hashtable<>();
   static{
+    fillConfig(config, JavaVersion.getVersion());
+  }
+
+  static void fillConfig(Map<String, String> config, int javaVersion) {
```
Comment (norrisjeremy) [inline src/main/java/com/jcraft/jsch/JSch.java:49]: What's the purpose of the `javaVersion` argument? It appears to be unused?

**Review by norrisjeremy — CHANGES_REQUESTED** [review]

File: src/main/java/com/jcraft/jsch/JSch.java (hunk 3 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/JSch.java:49]: It allows to test the different behavior in dependence of different java versions without the requirement to actually use that particular java version when executing the test

Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/JSch.java:49]: What's missing is a checkFillConfiig in JschTest but that's a big TODO I haven't had the time to attack

File: src/main/java/com/jcraft/jsch/Session.java (hunk 2 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/Session.java]: Changing the client version while being in an active session was possible before so we've found a possible bug here as far as I can see it. I've solved it now by recreating a member V_C that is filled with the bytes of the client version that is returned by `getClientVersion` at the time of connect.

File: src/main/java/com/jcraft/jsch/Session.java [hunk 4]
```diff
@@ -81,7 +81,8 @@
   private static final int PACKET_MAX_SIZE = 256 * 1024;
 
   private byte[] V_S;                                 // server version
-  private byte[] V_C=Util.str2byte("SSH-2.0-JSCH_"+JSch.VERSION); // client version
+  private byte[] V_C; // will be set during connect so that the client version
```
Comment (norrisjeremy) [inline src/main/java/com/jcraft/jsch/Session.java:84]: Why not simply do `VC=Util.str2byte(jsch.getClientVersion())` here?
Then we could omit all the other changes to this class.

File: src/main/java/com/jcraft/jsch/JSch.java (hunk 3 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/JSch.java:49]: ... and the actual use of the parameter was missing as well. Not sure what happened here, I've definitely used it when doing the implementation.

File: src/main/java/com/jcraft/jsch/Session.java (hunk 4 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/Session.java:84]: There might be a setClientVersion on Jsch-level after creating a session and before doing a connect that would get lost that way

Comment (norrisjeremy) [reply src/main/java/com/jcraft/jsch/Session.java:84]: I think it's easier to document and understand that calls to `jsch.setClientVersion()` have no impact to any previously created Session instances than it is to document and understand that `jsch.setClientVersion()` affects any previously created Session instance that has not yet execute `session.connect()`.

Basically: after you create a Session object, via `jsch.getSession()`, the only way to manipulate the client version string of said Session object should be via `session.setClientVersion()`, independent of any calls to `jsch.setClientVersion()` that occur after you have created the Session object.

File: src/main/java/com/jcraft/jsch/JSch.java (hunk 3 above)
Comment (norrisjeremy) [reply src/main/java/com/jcraft/jsch/JSch.java:49]: I'm still not sure I understand why we need a `javaVersion` argument instead of simply using `JavaVersion.getVersion()`?
The value returned by `JavaVersion.getVersion()` doesn't dynamically change at runtime.

Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/JSch.java:49]: I've committed JschTest showing the use of this parameter as part of a unit test. The fact that JavaVersion.getVersion doesn't change at runtime is the actual problem if you want to test behavior that depends on this value

File: src/test/java/com/jcraft/jsch/JSchTest.java [hunk 5]
```diff
@@ -64,6 +97,44 @@ void checkLoggerBehavior() throws Exception {
...
+      try {
+        System.setProperties(props);
+        
+        HashMap<String, String> map = new HashMap<>();
+        
+        JSch.fillConfig(map, 10);
+        assertEquals("com.jcraft.jsch.bc.XDH", map.get("xdh"), "check of xdh failed");
+        JSch.fillConfig(map, 11);
+        assertEquals("com.jcraft.jsch.jce.XDH", map.get("xdh"), "check of xdh failed");
+        JSch.fillConfig(map, 12);
```
Comment (norrisjeremy) [inline src/test/java/com/jcraft/jsch/JSchTest.java:116]: Do we need to test this for `12` since it is the same as `11`?

File: src/test/java/com/jcraft/jsch/JSchTest.java [hunk 6]
```diff
@@ -64,6 +97,44 @@ void checkLoggerBehavior() throws Exception {
...
+        
+        JSch.fillConfig(map, 14);
+        assertEquals("com.jcraft.jsch.bc.KeyPairGenEdDSA", map.get("keypairgen.eddsa"), "check of keypairgen.eddsa failed");
+        assertEquals("com.jcraft.jsch.bc.SignatureEd25519", map.get("ssh-ed25519"), "check of ssh-ed25519 failed");
+        assertEquals("com.jcraft.jsch.bc.SignatureEd448", map.get("ssh-ed448"), "check of ssh-ed448 failed");
+        JSch.fillConfig(map, 15);
+        assertEquals("com.jcraft.jsch.jce.KeyPairGenEdDSA", map.get("keypairgen.eddsa"), "check of keypairgen.eddsa failed");
+        assertEquals("com.jcraft.jsch.jce.SignatureEd25519", map.get("ssh-ed25519"), "check of ssh-ed25519 failed");
+        assertEquals("com.jcraft.jsch.jce.SignatureEd448", map.get("ssh-ed448"), "check of ssh-ed448 failed");
+        JSch.fillConfig(map, 16);
```
Comment (norrisjeremy) [inline src/test/java/com/jcraft/jsch/JSchTest.java:127]: Do we need to test this for `16` since it is the same as `15`?

File: src/test/java/com/jcraft/jsch/JSchTest.java (hunk 5 above)
Comment (kimmerin) [reply src/test/java/com/jcraft/jsch/JSchTest.java:116]: I'm always testing the edge cases to make sure that if somebody for some reason changes `>=` to `==`, the test fails.

File: src/test/java/com/jcraft/jsch/JSchTest.java (hunk 6 above)
Comment (kimmerin) [reply src/test/java/com/jcraft/jsch/JSchTest.java:127]: Same reason as with the test for xdh

File: src/test/java/com/jcraft/jsch/JSchTest.java [hunk 7]
```diff
@@ -1,17 +1,29 @@
...
 
 class JSchTest {
+    private Hashtable<String, String> orgConfig;
+    private Logger orgLogger;
 
     @BeforeEach
+    @SuppressWarnings("unchecked")
     void resetJsch() {
-        JSch.setLogger(null);
+      orgConfig = (Hashtable<String, String>) JSch.config.clone();
```
Comment (norrisjeremy) [inline src/test/java/com/jcraft/jsch/JSchTest.java:18]: How about simply doing `new Hashtable<>(JSch.config)`, to avoid the ugly cast and call to `clone()`?

Comment (kimmerin) [reply src/test/java/com/jcraft/jsch/JSchTest.java]: The "copy-constructor" creates shallow copies but with immutable values this isn't an issue here, so I've changed it as suggested.

File: src/main/java/com/jcraft/jsch/Session.java (hunk 4 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/Session.java:84]: Even if we implement it the way you suggested you need to either have a distinct member for the client version bytes used during a connection or check the state of the session and throw an exception if you try to set the version string after a connection has been established.

Otherwise you will run into the problem you've described before:

> If a user changed the client version string after a Session is connected and a rekeying
> operation occurred, their SSH connection would then break, because the client version
> is supposed to be idempotent.

Personally I don't think that this is hard to document and I can add a javadoc comment trying to do it to the method (I would have done so anyway, if the method didn't exist already).

File: src/test/java/com/jcraft/jsch/JSchTest.java (hunk 5 above)
Comment (kimmerin) [reply src/test/java/com/jcraft/jsch/JSchTest.java:115]: To `-v` that: `a >= b` can be seen as a shortcut for `a == b || a > b`. To have complete coverage over the conditions you need to implement three tests: one for `a < b`, one for `a == b` and one for `a > b`.

File: src/main/java/com/jcraft/jsch/Session.java (hunk 4 above)
Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/Session.java:84]: Added some javadoc that should explain the way this works.

Comment (norrisjeremy) [reply src/main/java/com/jcraft/jsch/Session.java:84]: I still disagree with this.
@mwiede what are your thoughts?

File: src/main/java/com/jcraft/jsch/Session.java [hunk 8]
```diff
@@ -2718,15 +2719,45 @@ public void setTimeout(int timeout) throws JSchException {
...
     return Util.byte2str(V_S);
   }
+  
+  /**
+   * Returns the client version string used during an active connection.
+   * If no connection is active <code>null</code> is returned.
+   * @return The used version string or <code>null</code> if no
+   * connection is active.
+   */
+  public String getUsedClientVersion() {
```
Comment (norrisjeremy) [inline src/main/java/com/jcraft/jsch/Session.java:2729]: I personally find this extremely convoluted and would really rather we not add this method.

Comment (kimmerin) [reply src/main/java/com/jcraft/jsch/Session.java]: Since two members are necessary to prevent the effect you've described I though, offering a get-method to allow access to these two members should be available. But I have no special feelings about this method, so I've removed it.

File: src/main/java/com/jcraft/jsch/Session.java (hunk 4 above)
Comment (norrisjeremy) [reply src/main/java/com/jcraft/jsch/Session.java:84]: @mwiede Any thoughts on this?
Should we just go with a simpler approach I suggested or the more complex approach implemented here?
